File: e2e/native/elixir/test/anvil_test.exs
defmodule AbiTypegenElixir.AnvilTest do
  use ExUnit.Case, async: false

  @moduletag :anvil
  @receipt_attempts 50
  @receipt_delay_ms 100
  @amount 340_282_366_920_938_463_463_374_607_431_768_211_456

  test "local signer mints 2^128, approves, reads, decodes events, and propagates reverts" do
    url = System.fetch_env!("ATG_RPC_URL")
    contract = System.fetch_env!("ATG_TOKEN_ADDRESS")
    private_key = System.fetch_env!("ATG_PRIVATE_KEY")
    chain_id = System.fetch_env!("ATG_CHAIN_ID") |> String.to_integer()
    rpc_opts = [url: url]

    {:ok, [owner]} = Ethers.Signer.Local.accounts(private_key: private_key)

    send_opts = [
      to: contract,
      from: owner,
      chain_id: chain_id,
      gas: 300_000,
      signer: Ethers.Signer.Local,
      signer_opts: [private_key: private_key],
      rpc_opts: rpc_opts
    ]

    {:ok, mint_hash} = Token.mint(owner, @amount) |> Ethers.send_transaction(send_opts)
    assert %{"status" => "0x1"} = await_receipt!(mint_hash, rpc_opts)

    assert {:ok, @amount} =
             Token.balance_of(owner) |> Ethers.call(to: contract, rpc_opts: rpc_opts)

    {:ok, approve_hash} = Token.approve(owner, @amount) |> Ethers.send_transaction(send_opts)
    assert %{"status" => "0x1"} = await_receipt!(approve_hash, rpc_opts)

    assert {:ok, @amount} =
             Token.allowance(owner, owner) |> Ethers.call(to: contract, rpc_opts: rpc_opts)

    assert {:ok, [%Ethers.Event{} = transfer]} =
             Ethers.get_logs(Token.EventFilters.transfer(nil, owner),
               address: contract,
               from_block: 0,
               to_block: "latest",
               rpc_opts: rpc_opts
             )

    assert transfer.data == [@amount]

    zero = "0x0000000000000000000000000000000000000000"

    assert {:error, %Token.Errors.InvalidRecipient{}} =
             Token.transfer(zero, 1) |> Ethers.call(to: contract, from: owner, rpc_opts: rpc_opts)
  end

  defp await_receipt!(hash, rpc_opts, attempts \\ @receipt_attempts)

  defp await_receipt!(hash, _rpc_opts, 0) do
    flunk("transaction receipt not available: #{hash}")
  end

  defp await_receipt!(hash, rpc_opts, attempts) do
    case Ethers.get_transaction_receipt(hash, rpc_opts: rpc_opts) do
      {:error, :transaction_receipt_not_found} ->
        Process.sleep(@receipt_delay_ms)
        await_receipt!(hash, rpc_opts, attempts - 1)

      {:ok, receipt} when is_map(receipt) ->
        receipt

      {:error, reason} ->
        flunk("receipt lookup failed for #{hash}: #{inspect(reason)}")
    end
  end
end

File: e2e/native/dart/test/anvil_test.dart
import 'dart:async';
import 'dart:io';

import 'package:abi_typegen_dart_e2e/generated/token.dart';
import 'package:test/test.dart';
import 'package:wallet/wallet.dart';
import 'package:web3dart/web3dart.dart';
import 'package:http/http.dart' as http;

void main() {
  final rpc = Platform.environment['ATG_RPC_URL'];
  final tokenAddress = Platform.environment['ATG_TOKEN_ADDRESS'];
  final privateKey = Platform.environment['ATG_PRIVATE_KEY'];
  final chainId = int.tryParse(Platform.environment['ATG_CHAIN_ID'] ?? '');
  if (rpc == null || tokenAddress == null || privateKey == null || chainId == null) {
    test('Anvil RPC binding test requires the supplied disposable-chain environment', () {}, skip: true);
    return;
  }

  test('signed mint and approve update state and decode the typed event', () async {
    final httpClient = http.Client();
    final client = Web3Client(rpc, httpClient);
    addTearDown(() async {
      client.dispose();
      httpClient.close();
    });
    final credentials = EthPrivateKey.fromHex(privateKey);
    final owner = credentials.address;
    final amount = BigInt.one << 128;
    final token = Token(EthereumAddress.fromHex(tokenAddress));

    final mintHash = await token.mint(
      client,
      credentials,
      chainId: chainId,
      to: owner,
      amount: amount,
    );
    final mintReceipt = await _receipt(client, mintHash);
    expect(mintReceipt.status, isTrue);
    expect(await token.balanceOf(client, value0: owner), amount);
    final transferLog = mintReceipt.logs.singleWhere(
      (log) => log.topics?.first == Token.transferTopic0,
    );
    final transfer = token.decodeTransferEvent(transferLog.topics!, transferLog.data!);
    expect(transfer.from, EthereumAddress.fromHex('0x0000000000000000000000000000000000000000'));
    expect(transfer.to, owner);
    expect(transfer.amount, amount);

    final spender = EthereumAddress.fromHex('0x0000000000000000000000000000000000000002');
    final approveHash = await token.approve(
      client,
      credentials,
      chainId: chainId,
      spender: spender,
      amount: BigInt.from(77),
    );
    expect((await _receipt(client, approveHash)).status, isTrue);
    expect(await token.allowance(client, value0: owner, value1: spender), BigInt.from(77));
    final logs = await token.getApprovalLogs(client, owner: owner, spender: spender);
    expect(logs, isNotEmpty);

    await expectLater(
      token.transfer(client, credentials, chainId: chainId, to: EthereumAddress.fromHex('0x0000000000000000000000000000000000000000'), amount: BigInt.one),
      throwsA(anything),
    );
  });
}

Future<TransactionReceipt> _receipt(Web3Client client, String hash) async {
  for (var attempt = 0; attempt < 50; attempt++) {
    final receipt = await client.getTransactionReceipt(hash);
    if (receipt != null) return receipt;
    await Future<void>.delayed(const Duration(milliseconds: 100));
  }
  throw TimeoutException('transaction receipt was not mined within 5 seconds: $hash');
}
